### PR TITLE
[ENG-2291] Adding a script for mass withdrawal for preprints.

### PR DESCRIPTION
**Pull request by corbinSanders**
## Purpose

We need a script to mass withdraw all preprints from one provider

## Changes

Adding the script. Adding tests. Modifying email template to include withdrawal justification.

## QA Notes

Please make verification statements inspired by your code and what your code touches.
- Verify Emails are sent correctly
- Verify Preprints are properly withdrawn

What are the areas of risk? Shouldn't be.

Any concerns/considerations/questions that development raised? N/A

## Documentation

N/A

## Side Effects

N/A

## DevOps Notes

DevOps will need to populate the user field in settings/defaults.py before uncommenting it for it to run.

## Ticket

https://openscience.atlassian.net/browse/ENG-2291

### CODE REVIEW

File: osf/management/commands/withdraw_all_preprints_from_provider.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        )
+        parser.add_argument(
+            '--comment',
+            dest='comment',
+            help='Comment to withdraw with'
+        )
+
+    def handle(self, *args, **options):
+        provider_id = options.get('provider_id', None)
+        dry_run = options.get('dry_run', False)
```
Comment (mfraezz) [inline osf/management/commands/withdraw_all_preprints_from_provider.py:84]: This script will generate emails, which currently prevents `dry_run` from actually being a dry run. Is it possible to conditionally override the mail settings to print to stdout instead, i.e. `mails.settings.USE_EMAIL = False`? If not, it seems like it shouldn't be an option at all.

File: website/templates/emails/withdrawal_request_granted.html.mako [hunk 2]
```diff
@@ -29,12 +29,12 @@
...
+            The ${document_type} has been removed from ${reviewable.provider.name}, but its metadata is still available: title of the withdrawn ${document_type}, its contributor list, abstract, tags, DOI, and reason for withdrawal (if provided). The moderator has provided the following justification: "${reviewable.withdrawal_justification}"
             <br>
         % else:
             ${requester.fullname} has withdrawn your ${document_type} <a href="${reviewable.absolute_url}">"${reviewable.title}"</a> from ${reviewable.provider.name}.
             <br>
-            The ${document_type} has been removed from ${reviewable.provider.name}, but its metadata is still available: title of the withdrawn ${document_type}, its contributor list, abstract, tags, DOI, and reason for withdrawal (if provided).
+            The ${document_type} has been removed from ${reviewable.provider.name}, but its metadata is still available: title of the withdrawn ${document_type}, its contributor list, abstract, tags, DOI, and reason for withdrawal (if provided). ${requester.fullname} provided the following justification: "${reviewable.withdrawal_justification}"
```
Comment (mfraezz) [inline website/templates/emails/withdrawal_request_granted.html.mako:37]: >its metadata is still available: [...] reason for withdrawal (if provided)

1) This implies it's possible for withdrawals to be "unjustified" -- should the added language be conditional on the presence of it?
2) Minor: Should "`and reason for withdrawal (if provided)`" be removed here? Seems duplicative

**Review by mfraezz — CHANGES_REQUESTED** [review]
Generally looks good, a couple small concerns. Pass complete :octocat:

File: osf/management/commands/withdraw_all_preprints_from_provider.py (hunk 1 above)
Comment (corbinSanders) [reply osf/management/commands/withdraw_all_preprints_from_provider.py]: Good point. It looks like suppressing emails might be too tricky, so I removed dry run.

File: website/templates/emails/withdrawal_request_granted.html.mako (hunk 2 above)
Comment (corbinSanders) [reply website/templates/emails/withdrawal_request_granted.html.mako]: Withdrawals should always have justifications AFAIK. To your second point, yes. Removed.

Comment (mfraezz) [reply website/templates/emails/withdrawal_request_granted.html.mako]: I just confirmed on staging blank justifications can be submitted.

File: osf/management/commands/withdraw_all_preprints_from_provider.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    if not comment and not saved_comment:
+        raise RuntimeError('Comment not provided!')
+
+    preprints = provider.preprints.filter(date_withdrawn=None)[:page_size]
+
+    preprints_withdrawn = 0
+
+    for preprint in preprints:
+        preprint_request = PreprintRequest(
```
Comment (mfraezz) [inline osf/management/commands/withdraw_all_preprints_from_provider.py:42]: Seeing `force_withdrawal` in the email template prompted me to dig a little. Calling `preprint.run_withdraw(user, comment)` is much simpler creating and walking through a request's approval, and also obfuscates the withdrawing user -- in my local test, it caused the `"A moderator has withdrawn your [...]"` email language, which seems preferable.

File: website/templates/emails/withdrawal_request_granted.html.mako (hunk 2 above)
Comment (corbinSanders) [reply website/templates/emails/withdrawal_request_granted.html.mako]: That upsets me.

**Review by mfraezz — APPROVED** [review]
💸

**Comment by mfraezz** [comment]
Cherry-picked in as https://github.com/CenterForOpenScience/osf.io/commit/113f7e0e14f2df5313b3b99580ee406ac04faa03 , consider this merged ⛏️

**Pull request closed without merging**